webhooks/unifiedstoragequota/validation/unifiedstoragequota_validator.go

```diff
@@ -0,0 +1,332 @@
+		Recorder:           record.New(mgr.GetEventRecorderFor(webhookNameLong)),
+		Logger:             ctx.Logger.WithName(webhookName),
+	}
+	// Initialize the webhook's decoder.
+	decoder := admission.NewDecoder(mgr.GetScheme())
+
+	handler := &RequestedCapacityHandler{
+		Client:         mgr.GetClient(),
+		Converter:      runtime.DefaultUnstructuredConverter,
+		Decoder:        decoder,
+		WebhookContext: webhookContext,
+	}
+	mgr.GetWebhookServer().Register(webhookPath, handler)
+
+	return nil
+}
+
+func (h *RequestedCapacityHandler) Handle(req admission.Request) CapacityResponse {
+	var (
+		obj, oldObj   *unstructured.Unstructured
+		handleRequest func(ctx *pkgctx.WebhookRequestContext) CapacityResponse
+	)
+
+	if req.Operation == v1.Create {
+		obj = &unstructured.Unstructured{}
+		if err := h.DecodeRaw(req.Object, obj); err != nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, err)}
+		}
+		handleRequest = h.HandleCreate
+	}
+	if req.Operation == v1.Update {
+		obj = &unstructured.Unstructured{}
+		if err := h.DecodeRaw(req.Object, obj); err != nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, err)}
+		}
+		oldObj = &unstructured.Unstructured{}
+		if err := h.DecodeRaw(req.OldObject, oldObj); err != nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, err)}
+		}
+		handleRequest = h.HandleUpdate
+	}
+
+	if obj == nil {
+		return CapacityResponse{Response: webhook.Allowed(string(req.Operation))}
+	}
+
+	webhookRequestContext := &pkgctx.WebhookRequestContext{
+		WebhookContext: h.WebhookContext,
+		Op:             req.Operation,
+		Obj:            obj,
+		OldObj:         oldObj,
+		UserInfo:       req.UserInfo,
+		Logger:         h.WebhookContext.Logger.WithName(obj.GetNamespace()).WithName(obj.GetName()),
+	}
+
+	return handleRequest(webhookRequestContext)
+}
+
+func (h *RequestedCapacityHandler) HandleCreate(ctx *pkgctx.WebhookRequestContext) CapacityResponse {
+	vm := &vmopv1.VirtualMachine{}
+	if err := h.Converter.FromUnstructured(ctx.Obj.UnstructuredContent(), vm); err != nil {
+		return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, err)}
+	}
+
+	scName := vm.Spec.StorageClass
+
+	sc := &storagev1.StorageClass{}
+	if err := h.Client.Get(ctx, client.ObjectKey{Name: scName}, sc); err != nil {
+		if apierrors.IsNotFound(err) {
+			return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+		}
+		return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+	}
+
+	var capacity *resource.Quantity
+
+	if vm.Status.Storage == nil || !conditions.IsTrue(vm, vmopv1.VirtualMachineConditionCreated) {
+		if vm.Spec.Image == nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, errors.New("vm.Spec.Image is required"))}
+		}
+		vmiName := vm.Spec.Image.Name
+		var imageStatus vmopv1.VirtualMachineImageStatus
+
+		switch vm.Spec.Image.Kind {
```

Comment:
Add a default case to return `http.StatusBadRequest` for other invalid image kinds?